hwy/contrib/sort/order-emulate-inl.h

```diff
@@ -0,0 +1,224 @@
+                                 T* HWY_RESTRICT /* buf */) const {
+    const RebindToSigned<D> di;
+    using VI = Vec<decltype(di)>;
+    VI key = LtBinKey(v);
+    VI min = MinOfLanes(di, key);
+    Mask<D> m = RebindMask(d, Eq(min, key));
+    return MaxOfLanes(d, IfThenElseZero(m, v));
+  }
+
+  template <class D>
+  HWY_INLINE Vec<D> LastOfLanes(D d, Vec<D> v,
+                                T* HWY_RESTRICT /* buf */) const {
+    const RebindToSigned<D> di;
+    using VI = Vec<decltype(di)>;
+    VI key = LtBinKey(v);
+    VI max = MaxOfLanes(di, key);
+    Mask<D> m = RebindMask(d, Eq(max, key));
+    return MaxOfLanes(d, IfThenElseZero(m, v));
+  }
+
+  template <class D>
+  HWY_INLINE Vec<D> FirstValue(D d) const {
+    return Set(d, BitCastScalar<T>(NegativeInfOrLowestValue<TF>()));
+  }
+
+  template <class D>
+  HWY_INLINE Vec<D> LastValue(D d) const {
+    return Set(d, BitCastScalar<T>(PositiveInfOrHighestValue<TF>()));
+  }
+
+  // Returns the next distinct smaller value unless already -inf.
+  template <class D, class V = Vec<D>>
+  HWY_INLINE V PrevValue(D, V v) const {
+    return NextSortValueBits<true>(v);
+  }
+
+  // Next representable value in total order by ±1 ULP, saturating at ±Inf.
+  //   IsDown = false → next larger
+  //   IsDown = true  → next smaller
+  template <bool IsDown, class V>
+  HWY_INLINE V NextSortValueBits(V u) const {
+    const DFromV<V> d;
+    using M = Mask<decltype(d)>;
+    constexpr T kSignBit = SignMask<T>();
+    constexpr T kBoundaryUp = SignMask<T>() - 1;
+    const V sign_bit = Set(d, kSignBit);
+    const V all1 = Set(d, T(~T(0)));
+    const V one = Set(d, T(1));
+    // Detect saturation at ±Inf
+    const M is_target_inf = Eq(u, IsDown ? FirstValue(d) : LastValue(d));
+    // Transform to monotonic space: flip sign for positives, invert for negatives
+    const M is_neg = TestBit(u, sign_bit);
+    const V key = Xor(u, IfThenElse(is_neg, all1, sign_bit));
+    // Boundary detection: +0/-0 swap needs a step of 2 instead of 1
+    const V boundary = Set(d, IsDown ? kSignBit : kBoundaryUp);
+    const M at_boundary = Eq(key, boundary);
+    // Step size: normally 1, but 2 at zero-boundary
+    const V step = Add(one, IfThenElseZero(at_boundary, one));
+    // Apply increment/decrement unless already at ±Inf
+    const V key2 = IfThenElse(is_target_inf, key,
+                              IsDown ? Sub(key, step) : Add(key, step));
+    // Transform back from monotonic space
+    const M neg_out = Lt(key2, sign_bit);
+    return Xor(key2, IfThenElse(neg_out, all1, sign_bit));
+  }
+};
+
+template <class Base>
+struct OrderEmulate<Base, SortDescending> : public OrderEmulate<Base, SortAscending> {
+  using _AscBase = OrderEmulate<Base, SortAscending>;
+  using _Asc = const _AscBase*;
+  using T = typename Base::LaneType;
+  using TF = typename Base::KeyType;
+
+  HWY_INLINE bool Compare1(const T* a, const T* b) const {
+    return reinterpret_cast<_Asc>(this)->Compare1(b, a);
```

Comment:
reinterpret_cast and inheritance seems questionable here.
We could instead use a typedef like you have (but no _ prefix because that is reserved by C++), and instead write return AscBase().Compare(d, b, a) - WDYT?

Reply:
Make sense, done